### PR TITLE
fix(pages): resync congrats screen after filtered deck rebuild

### LINKED ISSUE #21393: Filtered deck overview doesn't update properly after rebuilding
### Checked for duplicates?

- [x] This issue is not a duplicate

### Does it also happen in the desktop version?

- [x] This bug does not occur in the latest version of Anki Desktop

### What are the steps to reproduce this bug?

1. Make a filtered deck which successfully pulls some cards (i.e. is not empty).
2. Long press the deck and empty it.
3. Open the deck. You should see the congratulation screen.
4. Tap the button in the top right corner and rebuild the deck.

The deck overview still says "Congratulations...".

Image

### Expected behaviour

It should show the usual numbers and study button like this:

Image

### Debug info

```text
AnkiDroid Version = 2.25.0alpha2 (ad96da5a09ab3aaa3ddebb4928067c73168ff6c6)  
Backend Version = 0.1.64-anki25.09.2 (25.09.2 3890e12c9e48c028c3f12aa58cb64bd9f8895e30)  
Android Version = 13 (SDK 33)  
ProductFlavor = full  
Device Info = Xiaomi | POCO | mondrian | mondrian_global | 23013PC75G | qcom  
WebView Info = [com.google.android.webview | 787104603]: Mozilla/5.0 (Linux; Android 13; 23013PC75G Build/TKQ1.220905.001; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/150.0.7871.46 Mobile Safari/537.36  
ACRA UUID = 4afc3b3f-e93f-4a66-bc71-2b9476ef9eb0  
FSRS = 5.1.0 (Enabled: true)  
Crash Reports Enabled = true
```

### (Optional) Anything else you want to share?

The padding in the congratulations screen is too small.

### Research

- [x] I have checked the [manual](https://docs.ankidroid.org/) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by Ayush-Patel-56**
## Purpose / Description
The congrats screen for a filtered deck stays stuck on "Congratulations" after rebuilding, even when the rebuild pulls cards back in.

## Fixes
* Fixes #21393

## Approach
`CongratsPage` reloaded its own congrats WebView route on every `studyQueues` change without checking if the deck was actually still empty. `CongratsViewModel` now checks `sched.counts()` first and navigates to `StudyOptionsActivity` instead of reloading when the deck has cards due. Desktop's `Overview` already does this same check on every refresh, AnkiDroid's split-screen version just never ported it over.

## How Has This Been Tested?
Added `CongratsViewModelTest`, reproducing the repro steps from the issue (fill deck, empty it, rebuild). Ran the full unit test suite locally. Manually reproduced and verified the fix on a Pixel 7 emulator, API 33.

## Learning (optional, can help others)
Compared against desktop's `qt/aqt/overview.py` to see how it avoids this class of bug: it re-checks `sched._is_finished()` on every refresh instead of assuming the previous state still holds.

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by Ayush-Patel-56** [comment]
Hi @ZornHadNoChoice, can you please re-run the jobs ?

**Comment by ZornHadNoChoice** [comment]
I'm not sure I have the ability to do that (I don't even know what "jobs" are!).

**Comment by Ayush-Patel-56** [comment]
Open the failing test cases and on the top right corner you'll find an option to re-run jobs, so that the test cases runs again

**Comment by ZornHadNoChoice** [comment]
There's no such option

Screenshot_2026-07-29-11-06-02-196_com kiwibrowser browser-edit

**Comment by Ayush-Patel-56** [comment]
Oh I think you don't have an access to it, no worries any of the reviewer can do it. @criticalAY can you please help me with this ?

**Comment by BrayanDSO** [comment]
You don't need to ping a maintainer. Either wait or close and reopen the PR to trigger the jobs again

**Comment by Ayush-Patel-56** [comment]
the Lint Kotlin failure isn't from this PR, it's pre-existing on main. `allow_dangerous_js_api_title` and `allow_card_external_launch_title` in `10-preferences.xml` already exceed the 41-char limit from `PreferencesTitleMaxLengthAttr`, unmodified by this branch. this PR only touches `CongratsPage.kt` and adds the new test file. JUnit ubuntu run failed on a gradle dependency download timeout, unrelated too, passed on rerun.

**Comment by lukstbit** [comment]
A fix for the main branch lint failure was merged. Rebase on top of main to clear CI failures(if your PR doesn't introduce any).

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt [hunk 1]
```diff
@@ -265,10 +274,29 @@ class CongratsViewModel :
...
         }
     }
+
+    val congratsRefreshState = MutableSharedFlow<CongratsRefreshState>()
+
+    fun onStudyQueuesChanged() {
+        launchCatchingIO {
+            // desktop's Overview re-checks sched._is_finished() on every refresh instead of
+            // assuming the deck is still done - do the same before reloading the congrats page
+            val stillFinished = withCol { sched.counts().let { it.new + it.lrn + it.rev == 0 } }
```
Comment (sanjaysargam) [inline AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: I'd extract this as a function `isDeckFinished()` to improve readability

File: AnkiDroid/src/test/java/com/ichi2/anki/pages/CongratsViewModelTest.kt
Comment (sanjaysargam) [inline AnkiDroid/src/test/java/com/ichi2/anki/pages/CongratsViewModelTest.kt:1]: add a test for "still empty after rebuild"

**Review by sanjaysargam — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt (hunk 1 above)
Comment (Ayush-Patel-56) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: good call, extracted it

File: AnkiDroid/src/test/java/com/ichi2/anki/pages/CongratsViewModelTest.kt
Comment (Ayush-Patel-56) [reply AnkiDroid/src/test/java/com/ichi2/anki/pages/CongratsViewModelTest.kt:1]: added, thanks!

**Review by sanjaysargam — APPROVED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt [hunk 2]
```diff
@@ -123,6 +123,15 @@ class CongratsPage :
...
             }.launchIn(lifecycleScope)
 
+        viewModel.congratsRefreshState
+            .flowWithLifecycle(lifecycle)
+            .onEach { state ->
+                when (state) {
+                    CongratsRefreshState.ReloadPage -> webViewLayout.post { webViewLayout.reload() }
+                    CongratsRefreshState.DeckHasCardsToStudy -> openStudyOptionsAndFinish()
+                }
+            }.launchIn(lifecycleScope)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: I don't see how this would work. We wouldn't be subscribed when the event is emitted as the screen wouldn't be active/visible.

Comment (Ayush-Patel-56) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: emit() suspends until something's collecting, it doesn't drop the value. flowWithLifecycle just cancels/restarts the collector on stop/start, so a rebuild while we're stopped just parks until we resume. can add a test for that if you want it nailed down.

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: > emit() suspends until something's collecting

I don't believe this is the case, could you confirm empirically. In your test, you have a subscriber, driven by turbine.

We are calling `MutableSharedFlow.emit`: https://github.com/Kotlin/kotlinx.coroutines/blob/b157ff07152dc2adcfdb2e7e45ff2a1fd4dca0bd/kotlinx-coroutines-core/common/src/flow/SharedFlow.kt#L418-L421

We built the MutableStateFlow as: `MutableSharedFlow<CongratsRefreshState>()` which defines `replay = 0`

See the implementation of `tryEmitNoCollectorsLocked`: https://github.com/Kotlin/kotlinx.coroutines/blob/b157ff07152dc2adcfdb2e7e45ff2a1fd4dca0bd/kotlinx-coroutines-core/common/src/flow/SharedFlow.kt#L447-L455

```kt
     * If there are no subscribers, the buffer is not used.
     * Instead, the most recently emitted value is simply stored into
     * the replay cache if one was configured, displacing the older elements there,
     * or dropped if no replay cache was configured.
 ```

https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-mutable-shared-flow/emit.html

**Comment by david-allison** [comment]
Ok, I've looked into this more deeply:

This works, but only because `finish()` is called implicitly via:

```kt
            undoableOp<OpChangesWithId> {
                val safeAddUpdateResult = safeAddOrUpdateFilteredDeck(newFilterForUpdate)
                if (safeAddUpdateResult.isSuccess) {
                    state.update { DeckBuilt } // this calls through to `finish()`
```

Instead, could we explicitly perform a refresh on `STARTED`, rather than relying on an implicit ordering.

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt (hunk 2 above)
Comment (Ayush-Patel-56) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: you're right, with replay=0 and no active subscriber the emit is dropped, confirmed via the docs you linked. real fix: added an `onResume()` recheck in CongratsPage so it re-evaluates deck state every time the screen becomes visible again, not just when it catches the live event. wrote a test that rebuilds the deck while the fragment is stopped, then resumes it, failed without the fix, passes now.

**Comment by Ayush-Patel-56** [comment]
> Ok, I've looked into this more deeply:
> 
> This works, but only because `finish()` is called implicitly via:
> 
> ```kotlin
>             undoableOp<OpChangesWithId> {
>                 val safeAddUpdateResult = safeAddOrUpdateFilteredDeck(newFilterForUpdate)
>                 if (safeAddUpdateResult.isSuccess) {
>                     state.update { DeckBuilt } // this calls through to `finish()`
> ```
> 
> Instead, could we explicitly perform a refresh on `STARTED`, rather than relying on an implicit ordering.

plain `onStart()` has the same race in a different spot (fires before the collector resubscribes). moved the refresh inside the same `repeatOnLifecycle(STARTED)` block as the collector so ordering is guaranteed. pushed just now.

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt [hunk 3]
```diff
@@ -123,6 +126,24 @@ class CongratsPage :
...
+        // have been missed, so recheck on every STARTED entry instead of trusting only the
+        // opExecuted callback. The recheck is launched from inside repeatOnLifecycle so that
+        // congratsRefreshState always has an active collector before it's emitted to - emitting
+        // it from an unrelated callback (eg. onStart()) races the collector's own resubscription
+        // and can silently drop the event, since congratsRefreshState has no replay cache.
+        lifecycleScope.launch {
+            lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
+                launch { viewModel.onStudyQueuesChanged() }
+                viewModel.congratsRefreshState.collect { state ->
+                    when (state) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt:139]: Could you add a `Timber.d` here, I suspect this is going to be called twice. In the case of `reload()`, this will cause noticeable latency (and we'll want to resolve this).

Comment (Ayush-Patel-56) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt:142]: good catch, was scoped to the fragment's own lifecycle - `onViewCreated` can run more than once (webview gets torn down/recreated), so old collectors stuck around. switched to `viewLifecycleOwner`, also fixes the lint failure. added the `Timber.d`.

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt [hunk 4]
```diff
@@ -265,10 +292,31 @@ class CongratsViewModel :
...
+            // desktop's Overview re-checks sched._is_finished() on every refresh instead of
+            // assuming the deck is still done - do the same before reloading the congrats page
+            val stillFinished = isDeckFinished()
+            congratsRefreshState.emit(
+                if (stillFinished) CongratsRefreshState.ReloadPage else CongratsRefreshState.DeckHasCardsToStudy,
+            )
+        }
+    }
+
+    private suspend fun isDeckFinished(): Boolean = withCol { sched.counts().let { it.new + it.lrn + it.rev == 0 } }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: I would add `isEmpty` on the `Counts` object, or at least use `.count()`

https://github.com/ankidroid/Anki-Android/blob/21b0646b38b7dc480e28b79b9b42db7ef1bb5db9/libanki/src/main/java/com/ichi2/anki/libanki/sched/Counts.kt#L47-L51

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt [hunk 5]
```diff
@@ -65,8 +68,8 @@ class CongratsPage :
     ) {
         // typically due to 'day rollover'
         if (changes.studyQueues) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: I believe `unburyAndStudy()` will cause two navigations:

This doesn't perform a `== viewModel` for the handler (and the handler is not specified in undoableOp). 

```kt
    private suspend fun unburyAndStudy(mode: UnburyDeckRequest.Mode) {
        undoableOp {
            sched.unburyDeck(decks.getCurrentId(), mode)
        }
        unburyState.emit(UnburyState.OpenStudy)
    }
```

You probably want a `Timber.i` in `openStudyOptionsAndFinish` to confirm this.

----

**Sample check**

https://github.com/ankidroid/Anki-Android/blob/2a5a008a81546978805ce8af7993b383ecdbaa55/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt#L2394

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt:66]: add a TODO to move this to the ViewModel

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt [hunk 6]
```diff
@@ -123,6 +126,30 @@ class CongratsPage :
...
+        // opExecuted callback. The recheck is launched from inside repeatOnLifecycle so that
+        // congratsRefreshState always has an active collector before it's emitted to - emitting
+        // it from an unrelated callback (eg. onStart()) races the collector's own resubscription
+        // and can silently drop the event, since congratsRefreshState has no replay cache.
+        // Scoped to viewLifecycleOwner (not the fragment's own lifecycle) since onViewCreated can
+        // run more than once per fragment (PageFragment tears down/recreates the WebView) - using
+        // the fragment's lifecycle here would leave the previous view's loop running alongside it.
+        viewLifecycleOwner.lifecycleScope.launch {
+            viewLifecycleOwner.lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
+                launch { viewModel.onStudyQueuesChanged() }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: Do not use `onStudyQueuesChanged` here - add a ViewModel method which checks to see if there is an action to take.

* `onStudyQueuesChanged` should set a flag if `emit` does nothing, allowing us to defer the decision until we reach `STARTED` rather than dropping an event. Remove the flag when the event is consumed.
* Calling `onStudyQueuesChanged` here unconditionally adds latency - this performs an unnecessary `reload` on the first `STARTED` event (or if the user backgrounds the screen with no `studyQueues` change)
* Inside the ViewModel: unconditionally perform the `count()` check, as that's quick - `reload` is the latency-sensitive operation we only want to perform when the queues have changed.

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt (hunk 3 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt:142]: > onViewCreated can run more than once (webview gets torn down/recreated)

I don't believe this is the case, and the fix adds complexity to the code (with a verbose comment).

Please revert this (unless there's a case I haven't thought about): `SafeWebView` is the only 'refresh' we have, and this works on the `WebView` component.

Note: My 'double reload' was referring to the case where the `opChanges` occurred from a `rebuild` after the screen moved back to `STARTED` - I don't think I was correct here, but the line is useful as `reload` is expensive.

**Review by david-allison — CHANGES_REQUESTED** [review]
Thanks!! One more round I feel

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt (hunk 4 above)
Comment (Ayush-Patel-56) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: done, using `sched.counts().count() == 0` now

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt (hunk 5 above)
Comment (Ayush-Patel-56) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: `undoableOp` had no handler so this fragment's own `opExecuted` also fired and re-triggered navigation. passed handler = this in the viewmodel and filter it out in `opExecuted`. added `Timber.i` in `openStudyOptionsAndFinish` too.

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt
Comment (Ayush-Patel-56) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt:66]: added

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt (hunk 6 above)
Comment (Ayush-Patel-56) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: fair, the "called twice" case was actually the unbury handler issue above, not this. can't fully revert though - `viewLifecycleOwner` is what `UnsafeRepeatOnLifecycleDetector` requires here (that's the lint failure from before), but trimmed the comment down since the `onViewCreated-twice` reasoning wasn't the real justification.

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: This is still pending

**Comment by david-allison** [comment]
Thanks!! Last one 😁

File: AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt (hunk 6 above)
Comment (Ayush-Patel-56) [reply AnkiDroid/src/main/java/com/ichi2/anki/pages/CongratsPage.kt]: `onStudyQueuesChanged` sets a flag when `subscriptionCount` is 0 instead of emitting into the void, and a new `refreshIfNeeded()` runs on every STARTED: always does the cheap `count()` check, but only emits (reload/navigate) if the flag was set or the deck's no longer finished. pushed.

**Review by david-allison — APPROVED** [review]
LGTM, thanks!!

Fix the merge conflict and we can get this in

THANK YOU SO MUCH!!

**Review by david-allison — APPROVED** [review]
LGTM, I de-conflicted to get this in
